File: bracken/src/fern/semantic.ts
import type { FernEvent } from './events'
import {
  MAX_ADMINS,
  MAX_APP_NAME_BYTES,
  MAX_BAN_REASON_BYTES,
  MAX_CHANNEL_DESCRIPTION_BYTES,
  MAX_CHANNEL_ID_BYTES,
  MAX_CHANNEL_NAME_BYTES,
  MAX_GROUP_DESCRIPTION_BYTES,
  MAX_GROUP_NAME_BYTES,
  MAX_MESSAGE_TEXT_BYTES,
  MAX_NICKNAME_BYTES,
  MAX_REACTION_BYTES,
  MAX_RELAYS,
  MAX_RELAY_URL_BYTES,
} from './limits'
import { isValidEventId, isValidPubkey } from './utils'

export class SemanticValidationError extends Error {
  constructor(message: string) {
    super(message)
    this.name = 'SemanticValidationError'
  }
}

const textEncoder = new TextEncoder()

function byteLength(value: string): number {
  return textEncoder.encode(value).length
}

function stringField(
  value: unknown,
  field: string,
  { min = 0, max }: { min?: number; max: number },
): string {
  if (typeof value !== 'string') throw new SemanticValidationError(`${field} must be a string`)
  const size = byteLength(value)
  if (size < min) throw new SemanticValidationError(`${field} is too short`)
  if (size > max) throw new SemanticValidationError(`${field} exceeds maximum length`)
  return value
}

function pubkeyField(value: unknown, field: string): string {
  if (typeof value !== 'string' || !isValidPubkey(value)) {
    throw new SemanticValidationError(`${field} must be a pubkey`)
  }
  return value
}

function eventIdField(value: unknown, field: string): string {
  if (typeof value !== 'string' || !isValidEventId(value)) {
    throw new SemanticValidationError(`${field} must be an event id`)
  }
  return value
}

function integerField(value: unknown, field: string): number {
  if (typeof value !== 'number' || !Number.isInteger(value)) {
    throw new SemanticValidationError(`${field} must be an integer`)
  }
  return value
}

function only(content: Record<string, unknown>, allowed: string[]): void {
  const allowedSet = new Set(allowed)
  const extra = Object.keys(content).find((key) => !allowedSet.has(key))
  if (extra) throw new SemanticValidationError(`unexpected content field: ${extra}`)
}

function relayUrl(value: unknown): string {
  const url = stringField(value, 'relay', { min: 1, max: MAX_RELAY_URL_BYTES })
  let parsed: URL
  try {
    parsed = new URL(url)
  } catch {
    throw new SemanticValidationError('relay must be a ws:// or wss:// URL')
  }
  if (parsed.protocol !== 'ws:' && parsed.protocol !== 'wss:') {
    throw new SemanticValidationError('relay must be a ws:// or wss:// URL')
  }
  return url
}

function channelId(value: unknown, field = 'channel id'): string {
  return stringField(value, field, { min: 1, max: MAX_CHANNEL_ID_BYTES })
}

function validateChatChannel(raw: unknown): string {
  if (typeof raw !== 'object' || raw === null || Array.isArray(raw)) {
    throw new SemanticValidationError('chat channel must be an object')
  }
  const record = raw as Record<string, unknown>
  only(record, ['id', 'name', 'description', 'position'])
  const id = channelId(record['id'], 'channel.id')
  stringField(record['name'], 'channel.name', { min: 1, max: MAX_CHANNEL_NAME_BYTES })
  if ('description' in record) {
    stringField(record['description'], 'channel.description', { max: MAX_CHANNEL_DESCRIPTION_BYTES })
  }
  if ('position' in record) integerField(record['position'], 'channel.position')
  return id
}

export function validateEventSemantics(event: FernEvent): void {
  const c = event.content
  switch (event.type) {
    case 'genesis': {
      const required = ['name', 'description', 'public', 'founder', 'admins', 'relays', 'app']
      for (const field of required) {
        if (!(field in c)) throw new SemanticValidationError(`missing genesis field: ${field}`)
      }
      for (const key of Object.keys(c)) {
        if (!key.includes('.') && !required.includes(key)) {
          throw new SemanticValidationError(`unexpected genesis protocol field: ${key}`)
        }
      }
      stringField(c['name'], 'name', { min: 1, max: MAX_GROUP_NAME_BYTES })
      stringField(c['description'], 'description', { max: MAX_GROUP_DESCRIPTION_BYTES })
      if (typeof c['public'] !== 'boolean') throw new SemanticValidationError('public must be a boolean')
      const founder = pubkeyField(c['founder'], 'founder')
      if (founder !== event.author) throw new SemanticValidationError('founder must equal author')
      const admins = c['admins']
      if (!Array.isArray(admins) || admins.length < 1 || admins.length > MAX_ADMINS) {
        throw new SemanticValidationError('admins must be a non-empty bounded array')
      }
      for (const admin of admins) pubkeyField(admin, 'admin')
      if (!admins.includes(founder)) throw new SemanticValidationError('admins must include founder')
      const relays = c['relays']
      if (!Array.isArray(relays) || relays.length < 1 || relays.length > MAX_RELAYS) {
        throw new SemanticValidationError('relays must be a non-empty bounded array')
      }
      for (const relay of relays) relayUrl(relay)
      const app = stringField(c['app'], 'app', { min: 1, max: MAX_APP_NAME_BYTES })
      if (app === 'chat') {
        const channels = c['chat.channels']
        if (!Array.isArray(channels) || channels.length === 0) {
          throw new SemanticValidationError('chat.channels must be a non-empty array')
        }
        const ids = channels.map(validateChatChannel)
        if (!ids.includes('general')) throw new SemanticValidationError('chat.channels must include general')
        if ('chat.default_channel' in c) channelId(c['chat.default_channel'], 'chat.default_channel')
        if ('chat.system_channel' in c) channelId(c['chat.system_channel'], 'chat.system_channel')
      }
      return
    }
    case 'join':
    case 'leave':
      only(c, [])
      return
    case 'invite':
      only(c, ['invitee', 'role'])
      pubkeyField(c['invitee'], 'invitee')
      if (c['role'] !== 'member') throw new SemanticValidationError('role must be member')
      return
    case 'kick':
    case 'unban':
    case 'admin_add':
    case 'admin_remove':
      only(c, ['target'])
      pubkeyField(c['target'], 'target')
      return
    case 'ban': {
      only(c, ['target', 'until', 'reason'])
      pubkeyField(c['target'], 'target')
      if (c['until'] !== undefined && c['until'] !== null) {
        const until = integerField(c['until'], 'until')
        if (until <= 0) throw new SemanticValidationError('until must be positive')
      }
      stringField(c['reason'] ?? '', 'reason', { max: MAX_BAN_REASON_BYTES })
      return
    }
    case 'relay_update': {
      only(c, ['relays'])
      const relays = c['relays']
      if (!Array.isArray(relays) || relays.length < 1 || relays.length > MAX_RELAYS) {
        throw new SemanticValidationError('relays must be a non-empty bounded array')
      }
      for (const relay of relays) relayUrl(relay)
      return
    }
    case 'metadata_update':
      only(c, ['name', 'description'])
      if (!('name' in c) && !('description' in c)) {
        throw new SemanticValidationError('metadata_update must include a field')
      }
      if ('name' in c) stringField(c['name'], 'name', { min: 1, max: MAX_GROUP_NAME_BYTES })
      if ('description' in c) stringField(c['description'], 'description', { max: MAX_GROUP_DESCRIPTION_BYTES })
      return
    case 'chat.message':
      only(c, ['text', 'channel', 'reply_to'])
      stringField(c['text'], 'text', { min: 1, max: MAX_MESSAGE_TEXT_BYTES })
      channelId(c['channel'], 'channel')
      if (c['reply_to'] !== undefined && c['reply_to'] !== null) eventIdField(c['reply_to'], 'reply_to')
      return
    case 'chat.reaction':
      only(c, ['target', 'emoji'])
      eventIdField(c['target'], 'target')
      stringField(c['emoji'], 'emoji', { min: 1, max: MAX_REACTION_BYTES })
      return
    case 'chat.nickname_set':
      only(c, ['nickname'])
      stringField(c['nickname'], 'nickname', { min: 1, max: MAX_NICKNAME_BYTES })
      return
    case 'chat.channel_create':
      only(c, ['name', 'description', 'position'])
      stringField(c['name'], 'name', { min: 1, max: MAX_CHANNEL_NAME_BYTES })
      if ('description' in c) stringField(c['description'], 'description', { max: MAX_CHANNEL_DESCRIPTION_BYTES })
      if ('position' in c) integerField(c['position'], 'position')
      return
    case 'chat.channel_update':
      only(c, ['id', 'name', 'description', 'position'])
      channelId(c['id'], 'id')
      if (!('name' in c) && !('description' in c) && !('position' in c)) {
        throw new SemanticValidationError('channel_update must include an update')
      }
      if ('name' in c) stringField(c['name'], 'name', { min: 1, max: MAX_CHANNEL_NAME_BYTES })
      if ('description' in c) stringField(c['description'], 'description', { max: MAX_CHANNEL_DESCRIPTION_BYTES })
      if ('position' in c) integerField(c['position'], 'position')
      return
    case 'chat.channel_delete':
      only(c, ['id', 'name'])
      if (channelId(c['id'], 'id') === 'general') throw new SemanticValidationError('general channel cannot be deleted')
      if ('name' in c) stringField(c['name'], 'name', { min: 1, max: MAX_CHANNEL_NAME_BYTES })
      return
    case 'chat.settings_update':
      only(c, ['default_channel', 'system_channel'])
      if (!('default_channel' in c) && !('system_channel' in c)) {
        throw new SemanticValidationError('settings_update must include a field')
      }
      if ('default_channel' in c) channelId(c['default_channel'], 'default_channel')
      if ('system_channel' in c) channelId(c['system_channel'], 'system_channel')
      return
    default:
      throw new SemanticValidationError(`unknown event type: ${event.type}`)
  }
}

File: src/fern/events/semantic.py
"""Context-free semantic validation for core (protocol) events.

Core events are the bare types the protocol owns (membership, validator set,
group metadata). Application events (dotted types) are validated by their app
module, which reuses the generic field helpers exported here.

Validation here is context-free: membership, authorization, expected sequence
and certified-time bans depend on a particular finalized state and live in
``fern.bft.app`` / the app module.
"""
from __future__ import annotations

from fern.bft.validators import Validator, make_validator_set
from fern.crypto.encoding import is_valid_event_id_hex, is_valid_pubkey_hex
from fern.events.event import Event
from fern.events.limits import (
    MAX_APP_NAME_BYTES,
    MAX_BAN_REASON_BYTES,
    MAX_GROUP_DESCRIPTION_BYTES,
    MAX_GROUP_NAME_BYTES,
)
from fern.events.types import ProtocolTypes


class SemanticValidationError(ValueError):
    pass


# --- Generic content-field helpers (shared with app modules) -----------------


def byte_len(value: str) -> int:
    return len(value.encode("utf-8"))


def string_field(value: object, field: str, *, min_bytes: int = 0, max_bytes: int) -> str:
    if not isinstance(value, str):
        raise SemanticValidationError(f"{field} must be a string")
    size = byte_len(value)
    if size < min_bytes:
        raise SemanticValidationError(f"{field} is too short")
    if size > max_bytes:
        raise SemanticValidationError(f"{field} exceeds maximum length")
    return value


def pubkey_field(value: object, field: str) -> str:
    if not isinstance(value, str) or not is_valid_pubkey_hex(value):
        raise SemanticValidationError(f"{field} must be a pubkey")
    return value


def event_id_field(value: object, field: str) -> str:
    if not isinstance(value, str) or not is_valid_event_id_hex(value):
        raise SemanticValidationError(f"{field} must be an event id")
    return value


def int_field(value: object, field: str) -> int:
    if not isinstance(value, int) or isinstance(value, bool):
        raise SemanticValidationError(f"{field} must be an integer")
    return value


def only_fields(content: dict[str, object], allowed: set[str]) -> None:
    extra = set(content) - allowed
    if extra:
        raise SemanticValidationError(f"unexpected content field: {sorted(extra)[0]}")


# --- Core event validation ---------------------------------------------------


def _validate_validator_list(content: dict[str, object]) -> None:
    raw_validators = content.get("validators")
    if not isinstance(raw_validators, list):
        raise SemanticValidationError("validators must be an array")
    try:
        validators = [Validator.from_dict(raw) for raw in raw_validators if isinstance(raw, dict)]
        if len(validators) != len(raw_validators):
            raise ValueError("invalid validator entry")
        fault_tolerance = int_field(content["fault_tolerance"], "fault_tolerance")
        make_validator_set(validators, epoch=0, fault_tolerance=fault_tolerance)
    except ValueError as exc:
        raise SemanticValidationError(str(exc)) from exc


def validate_genesis_core(event: Event) -> None:
    """Validate the core (bare) fields of a genesis event.

    App-namespaced (dotted) fields are validated by the app module; this only
    checks that they belong to the declared app's namespace.
    """
    c = event.content
    required = {
        "chain_id",
        "name",
        "description",
        "public",
        "founder",
        "validators",
        "fault_tolerance",
        "app",
    }
    missing = required - set(c)
    if missing:
        raise SemanticValidationError(f"missing genesis field: {sorted(missing)[0]}")
    for key in c:
        if "." not in key and key not in required:
            raise SemanticValidationError(f"unexpected genesis protocol field: {key}")
    string_field(c["name"], "name", min_bytes=1, max_bytes=MAX_GROUP_NAME_BYTES)
    string_field(c["description"], "description", max_bytes=MAX_GROUP_DESCRIPTION_BYTES)
    if not isinstance(c["public"], bool):
        raise SemanticValidationError("public must be a boolean")
    founder = pubkey_field(c["founder"], "founder")
    if founder != event.author:
        raise SemanticValidationError("founder must equal author")
    chain_id = c["chain_id"]
    if not isinstance(chain_id, str) or not is_valid_event_id_hex(chain_id):
        raise SemanticValidationError("chain_id must be 64-char lowercase hex")
    _validate_validator_list(c)
    app = string_field(c["app"], "app", min_bytes=1, max_bytes=MAX_APP_NAME_BYTES)
    prefix = app + "."
    for key in c:
        if "." in key and not key.startswith(prefix):
            raise SemanticValidationError(f"genesis field for foreign app namespace: {key}")


def validate_core_event_semantics(event: Event) -> None:
    c = event.content
    t = event.type

    if t in (ProtocolTypes.JOIN, ProtocolTypes.LEAVE):
        only_fields(c, set())
    elif t == ProtocolTypes.INVITE:
        only_fields(c, {"invitee", "role"})
        pubkey_field(c.get("invitee"), "invitee")
        if c.get("role") != "member":
            raise SemanticValidationError("role must be member")
    elif t in (ProtocolTypes.KICK, ProtocolTypes.UNBAN):
        only_fields(c, {"target"})
        pubkey_field(c.get("target"), "target")
    elif t == ProtocolTypes.BAN:
        only_fields(c, {"target", "until", "reason"})
        pubkey_field(c.get("target"), "target")
        if c.get("until") is not None:
            until = int_field(c.get("until"), "until")
            if until <= 0:
                raise SemanticValidationError("until must be positive")
        string_field(c.get("reason", ""), "reason", max_bytes=MAX_BAN_REASON_BYTES)
    elif t == ProtocolTypes.VALIDATOR_UPDATE:
        only_fields(c, {"validators", "fault_tolerance", "readiness"})
        _validate_validator_list(c)
        if not isinstance(c.get("readiness"), list):
            raise SemanticValidationError("readiness must be an array")
    elif t == ProtocolTypes.METADATA_UPDATE:
        only_fields(c, {"name", "description"})
        if "name" not in c and "description" not in c:
            raise SemanticValidationError("metadata_update must include a field")
        if "name" in c:
            string_field(c["name"], "name", min_bytes=1, max_bytes=MAX_GROUP_NAME_BYTES)
        if "description" in c:
            string_field(c["description"], "description", max_bytes=MAX_GROUP_DESCRIPTION_BYTES)
    else:
        raise SemanticValidationError(f"unknown core event type: {t}")


__all__ = [
    "SemanticValidationError",
    "byte_len",
    "event_id_field",
    "int_field",
    "only_fields",
    "pubkey_field",
    "string_field",
    "validate_core_event_semantics",
    "validate_genesis_core",
]
